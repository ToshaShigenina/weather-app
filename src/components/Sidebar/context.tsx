import React from 'react';

export type SidebarContextType = {
	isOpenSearch: boolean,
	toggleSearch: (isOpen: boolean) => void,
}
export const SidebarContext = React.createContext<SidebarContextType | null>(null);
export const SidebarContextProvider: React.FC<React.HTMLAttributes<HTMLDivElement>> = ({
	children
}) => {
	const [isOpenSearch, setIsOpenSearch] = React.useState(false);
	const ctx: SidebarContextType = {
		isOpenSearch,
		toggleSearch: setIsOpenSearch
	};

	return (
		<SidebarContext.Provider value={ ctx }>
			{ children }
		</SidebarContext.Provider>
	)
};