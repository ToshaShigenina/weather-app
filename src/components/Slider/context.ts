import React from "react";

export type SliderContextType = {
	width: number
	margin: number
	sliderKeys: symbol[]
	addSliderKey?: (key: symbol) => void
}

export const SliderContext = React.createContext<SliderContextType | null>(null);
export const SliderContextProvider = SliderContext.Provider;
