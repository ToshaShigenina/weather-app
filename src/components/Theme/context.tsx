import React from 'react';

export type ThemeNameType = 'dark' | 'light'
export type ThemeContextType = {
	theme: ThemeNameType,
	setTheme: (theme: ThemeNameType) => void
}
const getInitialTheme: () => ThemeNameType = () => {
	if (window) {
		const initialTheme = window.localStorage.getItem('theme');
		if (initialTheme) {
			return initialTheme as ThemeNameType;
		}
		const userMedia = window.matchMedia('(prefers-color-scheme: dark)');
		if (userMedia && userMedia.matches) {
			return 'dark';
		}
	}
	return 'light';
};
const setRawTheme = (theme: ThemeNameType) => {
	localStorage.setItem('theme', theme);
	if (theme === 'dark') {
		document.body.classList.add('theme-dark');
	} else {
		document.body.classList.remove('theme-dark');
	}
};

export const ThemeContext = React.createContext<ThemeContextType | null>(null);
export const ThemeContextProvider: React.FC<React.HTMLAttributes<HTMLDivElement>> = ({ children }) => {
	const [theme, setTheme] = React.useState(getInitialTheme);
	const ctx: ThemeContextType = {
		theme,
		setTheme
	};
	React.useEffect(() => {
		setRawTheme(theme);
	});
	React.useEffect(() => {
		setRawTheme(theme);
	}, [theme]);

	return (
		<ThemeContext.Provider value={ ctx }>
			{ children }
		</ThemeContext.Provider>
	);
};