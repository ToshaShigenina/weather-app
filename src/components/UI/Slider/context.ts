import { createContext } from "react";

export type SliderContextType = {
	width: number
	margin: number
	sliderKeys: symbol[]
	addSliderKey?: (key: symbol) => void
}

export const SliderContext = createContext<SliderContextType | null>(null);
export const SliderContextProvider = SliderContext.Provider;
