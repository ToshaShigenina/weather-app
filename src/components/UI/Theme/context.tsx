import React, { FC, HTMLAttributes, useEffect, createContext } from 'react';
import { useLocalStorage } from "../../../hooks";

export type ThemeNameType = 'dark' | 'light'
export type ThemeContextType = {
	theme: ThemeNameType | string,
	setTheme: (theme: ThemeNameType) => void
}

export const THEME_KEY = 'theme';
export const THEME_CLASS_NAME = 'theme-dark';

const getInitialTheme = (): ThemeNameType => {
	const userMedia = typeof matchMedia === 'function' && matchMedia('(prefers-color-scheme: dark)');
	return (userMedia && userMedia.matches) ? 'dark' : 'light';
};
const changeThemeClass = (theme: ThemeNameType) => {
	if (theme === 'dark') {
		document.body.classList.add(THEME_CLASS_NAME);
	} else {
		document.body.classList.remove(THEME_CLASS_NAME);
	}
};

export const ThemeContext = createContext<ThemeContextType | null>(null);
export const ThemeContextProvider: FC<HTMLAttributes<HTMLDivElement>> = ({ children }) => {
	const [theme, setTheme] = useLocalStorage<ThemeNameType>(THEME_KEY, getInitialTheme());
	const ctx: ThemeContextType = { theme, setTheme };
	const changeTheme = (): void => {
		setTheme(theme as ThemeNameType);
		changeThemeClass(theme as ThemeNameType);
	};
	useEffect(changeTheme, []);
	useEffect(changeTheme, [theme]);

	return (
		<ThemeContext.Provider value={ ctx }>
			{ children }
		</ThemeContext.Provider>
	);
};