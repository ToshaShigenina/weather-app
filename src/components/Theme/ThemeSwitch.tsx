import React from 'react';
import cn from 'classnames';
import { Icon } from '../Icon';
import { ThemeContext, type ThemeNameType } from './context';

import './css/style.css';

const ThemeSwitch: React.FC<React.HTMLAttributes<HTMLDivElement>> = ({ className, ...otherProps }) => {
    const ctx = React.useContext(ThemeContext);
    const theme: ThemeNameType = ctx?.theme || 'light';

    const onToggleTheme = () => {
        if (ctx) {
            const newTheme = theme === 'dark' ? 'light' : 'dark';
            ctx.setTheme(newTheme)
        }
    };

    return (
        <div className={ cn(className, 'theme') } { ...otherProps }>
            <label className="theme__switch"
                htmlFor="theme" >
                <input id="theme"
                    type="checkbox"
                    className="theme__checkbox"
                    checked={ theme === 'dark' }
                    onChange={ onToggleTheme } />
                <div className="theme__switch-item">
                    <Icon type="switch" className="theme__icon" />
                </div>
            </label>
        </div>
    );
};

export default ThemeSwitch;