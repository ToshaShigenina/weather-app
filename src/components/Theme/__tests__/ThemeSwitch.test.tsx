import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { mountTest, localStorageMock, mathcMediaMock } from '../../../tests';

import ThemeSwitch from '../ThemeSwitch';
import { ThemeContextProvider, type ThemeNameType } from '../context';

const testId = 'switch';

describe('ThemeSwitch', () => {
	beforeAll(() => {
		Object.defineProperty(
			window,
			'matchMedia',
			mathcMediaMock(() => {
				return {
					matches: false,
				}
			})
		);
	})

	mountTest(ThemeSwitch);

	it('отрабатывает изменение темы', () => {

		const { baseElement } = render(
			<ThemeContextProvider>
				<ThemeSwitch data-testid={ testId } />
			</ThemeContextProvider>
		);

		expect(screen.getByTestId(testId)).toMatchSnapshot();
		expect(baseElement).toMatchSnapshot();
		expect(baseElement).not.toHaveClass('theme-dark');
		fireEvent.click(screen.getByRole('checkbox'));
		expect(baseElement).toHaveClass('theme-dark');
		expect(baseElement).toMatchSnapshot();
	});
	it('сразу устанавливается темная тема из темы браузера', () => {
		Object.defineProperty(
			window,
			'matchMedia',
			mathcMediaMock(() => {
				return {
					matches: true,
				}
			})
		);

		const { baseElement } = render(
			<ThemeContextProvider>
				<ThemeSwitch />
			</ThemeContextProvider>
		);

		expect(baseElement).toMatchSnapshot();
		expect(baseElement).toHaveClass('theme-dark');
		fireEvent.click(screen.getByRole('checkbox'));
		expect(baseElement).not.toHaveClass('theme-dark');
		expect(baseElement).toMatchSnapshot();
	});
	it('устанавливается светлая тема из localStorage', () => {
		Object.defineProperty(window, 'localStorage', {
			value: localStorageMock
		});

		const { baseElement } = render(
			<ThemeContextProvider>
				<ThemeSwitch />
			</ThemeContextProvider>
		);

		expect(baseElement).not.toHaveClass('theme-dark');
		expect(baseElement).toMatchSnapshot();
	});
});