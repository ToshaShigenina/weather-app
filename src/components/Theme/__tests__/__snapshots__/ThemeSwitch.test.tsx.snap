// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitch отрабатывает изменение темы 1`] = `
<div
  class="theme"
  data-testid="switch"
>
  <label
    class="theme__switch"
    for="theme"
  >
    <input
      class="theme__checkbox"
      id="theme"
      type="checkbox"
    />
    <div
      class="theme__switch-item"
    >
      <svg
        class="theme__icon svg-icon"
        height="24"
        width="24"
      >
        <use
          xlink:href="#switch"
        />
      </svg>
    </div>
  </label>
</div>
`;

exports[`ThemeSwitch отрабатывает изменение темы 2`] = `
<body>
  <div>
    <div
      class="theme"
      data-testid="switch"
    >
      <label
        class="theme__switch"
        for="theme"
      >
        <input
          class="theme__checkbox"
          id="theme"
          type="checkbox"
        />
        <div
          class="theme__switch-item"
        >
          <svg
            class="theme__icon svg-icon"
            height="24"
            width="24"
          >
            <use
              xlink:href="#switch"
            />
          </svg>
        </div>
      </label>
    </div>
  </div>
</body>
`;

exports[`ThemeSwitch отрабатывает изменение темы 3`] = `
<body
  class="theme-dark"
>
  <div>
    <div
      class="theme"
      data-testid="switch"
    >
      <label
        class="theme__switch"
        for="theme"
      >
        <input
          class="theme__checkbox"
          id="theme"
          type="checkbox"
        />
        <div
          class="theme__switch-item"
        >
          <svg
            class="theme__icon svg-icon"
            height="24"
            width="24"
          >
            <use
              xlink:href="#switch"
            />
          </svg>
        </div>
      </label>
    </div>
  </div>
</body>
`;

exports[`ThemeSwitch сразу устанавливается темная тема из темы браузера 1`] = `
<body
  class="theme-dark"
>
  <div>
    <div
      class="theme"
    >
      <label
        class="theme__switch"
        for="theme"
      >
        <input
          checked=""
          class="theme__checkbox"
          id="theme"
          type="checkbox"
        />
        <div
          class="theme__switch-item"
        >
          <svg
            class="theme__icon svg-icon"
            height="24"
            width="24"
          >
            <use
              xlink:href="#switch"
            />
          </svg>
        </div>
      </label>
    </div>
  </div>
</body>
`;

exports[`ThemeSwitch сразу устанавливается темная тема из темы браузера 2`] = `
<body
  class=""
>
  <div>
    <div
      class="theme"
    >
      <label
        class="theme__switch"
        for="theme"
      >
        <input
          checked=""
          class="theme__checkbox"
          id="theme"
          type="checkbox"
        />
        <div
          class="theme__switch-item"
        >
          <svg
            class="theme__icon svg-icon"
            height="24"
            width="24"
          >
            <use
              xlink:href="#switch"
            />
          </svg>
        </div>
      </label>
    </div>
  </div>
</body>
`;

exports[`ThemeSwitch устанавливается светлая тема из localStorage 1`] = `
<body
  class=""
>
  <div>
    <div
      class="theme"
    >
      <label
        class="theme__switch"
        for="theme"
      >
        <input
          class="theme__checkbox"
          id="theme"
          type="checkbox"
        />
        <div
          class="theme__switch-item"
        >
          <svg
            class="theme__icon svg-icon"
            height="24"
            width="24"
          >
            <use
              xlink:href="#switch"
            />
          </svg>
        </div>
      </label>
    </div>
  </div>
</body>
`;
